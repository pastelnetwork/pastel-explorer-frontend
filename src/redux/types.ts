import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';
import { RootState } from '@redux/store';

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export type AppThunkDispatch = ThunkDispatch<RootState, unknown, Action<string>>;
