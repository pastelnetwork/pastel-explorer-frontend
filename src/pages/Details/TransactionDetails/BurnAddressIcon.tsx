import Tooltip from '@mui/material/Tooltip';

import { translateDropdown } from '@utils/helpers/i18n';
import databaseIcon from '@assets/images/database.png';
import * as Styles from './TransactionDetails.styles';

interface IBurnAddressIcon {
  type: string;
}

const BurnAddressIcon: React.FC<IBurnAddressIcon> = ({ type }) => {
  if (type !== 'storage') {
    return null;
  }

  return (
    <Tooltip title={translateDropdown('pages.transactionDetails.burnAddress')}>
      <Styles.FireWrapper>
        <img src={databaseIcon} alt={translateDropdown('pages.transactionDetails.burnAddress')} />
      </Styles.FireWrapper>
    </Tooltip>
  );
};

export default BurnAddressIcon;
