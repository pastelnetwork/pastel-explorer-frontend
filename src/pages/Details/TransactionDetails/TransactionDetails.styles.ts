import styled from 'styled-components';

import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import { Alert as MuiAlert } from '@material-ui/lab';
import { Grid, Toolbar } from '@material-ui/core';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    appBar: {
      position: 'relative',
    },
    title: {
      marginLeft: theme.spacing(2),
      flex: 1,
    },
  }),
);

export const Alert = styled(MuiAlert)`
  justify-content: center;
`;

export const TransactionDesc = styled(Grid)`
  cursor: pointer;
`;

export const TransactionRawData = styled.pre`
  max-width: 100%;
  overflow: auto;
`;

export const TransactionRawDataToolbar = styled(Toolbar)`
  background-color: ${props => props.theme.sidebar.background.default};
`;
