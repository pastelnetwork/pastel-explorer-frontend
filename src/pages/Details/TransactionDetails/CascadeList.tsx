import Grid from '@material-ui/core/Grid';
import Box from '@material-ui/core/Box';

import { TCascadeRequests } from '@utils/types/ITransactions';
import RouterLink from '@components/RouterLink/RouterLink';
import * as ROUTES from '@utils/constants/routes';

import * as TableStyles from '@components/Table/Table.styles';
import * as TicketStyles from '@components/Ticket/Ticket.styles';
import * as BlockStyles from '@pages/Details/BlockDetails/BlockDetails.styles';
import * as Styles from './TransactionDetails.styles';

interface ICascadeList {
  data: TCascadeRequests[];
}

interface ICascadeItem {
  cascade: TCascadeRequests;
}

const CascadeItem: React.FC<ICascadeItem> = ({ cascade }) => {
  return (
    <>
      <Grid container spacing={3}>
        <Grid item xs={4} sm={2}>
          <TicketStyles.TicketTitle>Cascade ID:</TicketStyles.TicketTitle>
        </Grid>
        <Grid item xs={8} sm={10}>
          <TicketStyles.TicketContent>
            {cascade.cascadeId.indexOf('nocascade') === -1 ? (
              <RouterLink
                route={`${ROUTES.CASCADE_DETAILS}/${cascade.cascadeId}`}
                value={cascade.cascadeId}
                title={cascade.cascadeId}
                className="address-link"
              />
            ) : (
              <RouterLink
                route={`${ROUTES.CASCADE_DETAILS}/${cascade.cascadeId}`}
                value={cascade.cascadeId}
                title={cascade.cascadeId}
                className="address-link"
              />
            )}
          </TicketStyles.TicketContent>
        </Grid>
      </Grid>
    </>
  );
};

const CascadeList: React.FC<ICascadeList> = ({ data }) => {
  return (
    <BlockStyles.GridStyle item>
      <TableStyles.BlockWrapper className="mb-0">
        <TableStyles.BlockTitle>Cascades</TableStyles.BlockTitle>
        <Box className="custom-table tickets-table">
          {data.map(cascade => (
            <Styles.GridStyle item key={cascade.cascadeId} className="table__row">
              <CascadeItem cascade={cascade} />
            </Styles.GridStyle>
          ))}
        </Box>
      </TableStyles.BlockWrapper>
    </BlockStyles.GridStyle>
  );
};

export default CascadeList;
