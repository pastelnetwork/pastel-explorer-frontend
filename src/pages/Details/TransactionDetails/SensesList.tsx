import Grid from '@material-ui/core/Grid';
import Box from '@material-ui/core/Box';

import { TSenseRequests } from '@utils/types/ITransactions';
import RouterLink from '@components/RouterLink/RouterLink';
import * as ROUTES from '@utils/constants/routes';

import * as TableStyles from '@components/Table/Table.styles';
import * as TicketStyles from '@components/Ticket/Ticket.styles';
import * as BlockStyles from '@pages/Details/BlockDetails/BlockDetails.styles';
import * as Styles from './TransactionDetails.styles';

interface ISensesList {
  data: TSenseRequests[];
}

interface ISenseItem {
  sense: TSenseRequests;
}

const SenseItem: React.FC<ISenseItem> = ({ sense }) => {
  return (
    <>
      <Grid container spacing={3}>
        <Grid item xs={4} sm={3} className="max-w-355">
          <TicketStyles.TicketTitle>Image Hash:</TicketStyles.TicketTitle>
        </Grid>
        <Grid item xs={8} sm={9}>
          <TicketStyles.TicketContent>
            {sense.imageFileHash.indexOf('nosense') === -1 ? (
              <RouterLink
                route={`${ROUTES.SENSE_DETAILS}?txid=${sense.transactionHash}&hash=${sense.imageFileHash}`}
                value={sense.imageFileHash}
                title={sense.imageFileHash}
                className="address-link"
              />
            ) : (
              ''
            )}
          </TicketStyles.TicketContent>
        </Grid>
      </Grid>
      <Grid container spacing={3}>
        <Grid item xs={4} sm={3} className="max-w-355">
          <TicketStyles.TicketTitle>Sense Version:</TicketStyles.TicketTitle>
        </Grid>
        <Grid item xs={8} sm={9}>
          <TicketStyles.TicketContent>
            {sense.imageFileHash.indexOf('nosense') === -1 ? sense.dupeDetectionSystemVersion : ''}
          </TicketStyles.TicketContent>
        </Grid>
      </Grid>
    </>
  );
};

const SensesList: React.FC<ISensesList> = ({ data }) => {
  return (
    <BlockStyles.GridStyle item>
      <TableStyles.BlockWrapper className="mb-0">
        <TableStyles.BlockTitle>Senses</TableStyles.BlockTitle>
        <Box className="custom-table tickets-table">
          {data.map(sense => (
            <Styles.GridStyle item key={sense.imageFileHash} className="table__row">
              <SenseItem sense={sense} />
            </Styles.GridStyle>
          ))}
        </Box>
      </TableStyles.BlockWrapper>
    </BlockStyles.GridStyle>
  );
};

export default SensesList;
