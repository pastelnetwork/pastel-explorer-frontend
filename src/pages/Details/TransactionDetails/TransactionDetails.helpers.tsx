import { Grid } from '@material-ui/core';
import { Alert, AlertTitle } from '@material-ui/lab';

import { HeaderType } from '@components/Table/Table';

import { ITransactionDetails } from '@utils/types/ITransactions';
import { formattedDate } from '@utils/helpers/date/date';
import { formatNumber } from '@utils/helpers/formatNumbers/formatNumbers';

import * as Styles from './TransactionDetails.styles';

const BLOCK_CONFIRMED_NUMBER = 6;

export const transactionHeaders: Array<HeaderType> = [
  { id: 1, header: 'Confirmations' },
  { id: 2, header: 'Block Hash' },
  { id: 3, header: 'Timestamp' },
];

export const addressHeaders: Array<HeaderType> = [
  { id: 1, header: 'Address' },
  { id: 2, header: 'Amount (PSL)' },
  { id: 3, header: 'Amount (USD)' },
];

export const generateTableTitle = (transactionData: ITransactionDetails) => (
  <Alert severity="info">
    <AlertTitle style={{ wordBreak: 'break-word' }}>PSL ID: {transactionData.id}</AlertTitle>
    {`This transaction was first broadcast to the PSL network on 
      ${formattedDate(transactionData.timestamp)}. 
      The transaction is currently 
      ${
        transactionData.block.confirmations > BLOCK_CONFIRMED_NUMBER ? 'confirmed' : 'unconfirmed'
      } by the network.
      At the time of this transaction 
      ${formatNumber(transactionData.totalAmount, { decimalsLength: 2 })} 
      PSL was sent.`}
  </Alert>
);

export const generateCoinbaseInfo = (info: number) => (
  <Alert severity="info">
    <AlertTitle>New coins ({formatNumber(info, { decimalsLength: 2 })})</AlertTitle>
  </Alert>
);

export const generateNonStandardTransactionInfo = () => (
  <Grid item xs={12}>
    <Styles.Alert severity="info">
      <AlertTitle>NONSTANDARD TX</AlertTitle>
    </Styles.Alert>
  </Grid>
);
