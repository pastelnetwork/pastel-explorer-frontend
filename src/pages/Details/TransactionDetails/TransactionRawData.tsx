import * as React from 'react';
import { Typography, Dialog, AppBar, IconButton, Slide } from '@material-ui/core';
import { TransitionProps } from '@material-ui/core/transitions';
import CloseIcon from '@material-ui/icons/Close';

import { ITicket, TSenseRequests } from '@utils/types/ITransactions';
import { translate } from '@utils/helpers/i18n';

import * as Styles from './TransactionDetails.styles';

const Transition = React.forwardRef(function Transition(
  props: TransitionProps,
  ref: React.Ref<unknown>,
) {
  return <Slide direction="up" ref={ref} {...props} />;
});

interface ITransactionRawDataProps {
  rawData: string;
  open: boolean;
  toggleOpen: () => void;
  tickets: ITicket[];
  senses: TSenseRequests[];
}

const TransactionRawData: React.FC<ITransactionRawDataProps> = ({
  rawData,
  open,
  toggleOpen,
  tickets,
  senses,
}) => {
  const classes = Styles.useStyles();

  const getTickets = () => {
    return tickets.map(ticket => ({
      ...ticket.data.ticket,
      txid: ticket.transactionHash,
    }));
  };

  const getSenses = () => {
    return senses.map(sense => {
      const parseSenseData = JSON.parse(sense.rawData);
      return {
        ...parseSenseData,
        raw_dd_service_data_json: JSON.parse(parseSenseData.raw_dd_service_data_json),
      };
    });
  };

  return (
    <Dialog fullScreen open={open} onClose={toggleOpen} TransitionComponent={Transition}>
      <AppBar className={classes.appBar}>
        <Styles.TransactionRawDataToolbar>
          <IconButton edge="start" color="inherit" onClick={toggleOpen} aria-label="close">
            <CloseIcon />
          </IconButton>
          <Typography variant="h6" className={classes.title}>
            {translate('pages.transactionDetails.transactionRawData')}
          </Typography>
        </Styles.TransactionRawDataToolbar>
      </AppBar>
      {rawData ? (
        <Styles.TransactionRawData>
          <code>
            {JSON.stringify(
              {
                ...JSON.parse(rawData),
                tickets: tickets.length ? getTickets() : undefined,
                senses: senses.length ? getSenses() : undefined,
              },
              null,
              2,
            )}
          </code>
        </Styles.TransactionRawData>
      ) : (
        translate('common.noData')
      )}
    </Dialog>
  );
};

export default TransactionRawData;
