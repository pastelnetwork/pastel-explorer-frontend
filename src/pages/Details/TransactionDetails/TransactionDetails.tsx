import * as React from 'react';
import { useParams, Redirect } from 'react-router-dom';

import { Grid } from '@material-ui/core';

import Header from '@components/Header/Header';
import RouterLink from '@components/RouterLink/RouterLink';
import Table, { RowsProps } from '@components/Table/Table';
import CopyButton from '@components/CopyButton/CopyButton';

import { useFetch } from '@utils/helpers/useFetch/useFetch';
import { formatNumber } from '@utils/helpers/formatNumbers/formatNumbers';
import * as ROUTES from '@utils/constants/routes';
import * as URLS from '@utils/constants/urls';
import { formattedDate } from '@utils/helpers/date/date';
import { TransactionEvent, DirectionType, ITransactionDetails } from '@utils/types/ITransactions';
import { ISummary } from '@utils/types/ISummary';
import { useSortData } from '@utils/hooks';

import * as Styles from './TransactionDetails.styles';
import {
  addressHeaders,
  transactionHeaders,
  generateTableTitle,
  generateNonStandardTransactionInfo,
  generateCoinbaseInfo,
} from './TransactionDetails.helpers';
import TransactionRawData from './TransactionRawData';

interface ParamTypes {
  id: string;
}

const TransactionDetails = () => {
  const { id } = useParams<ParamTypes>();
  const [transaction, setTransaction] = React.useState<ITransactionDetails | null>(null);
  const [exchangeRate, setExchangeRate] = React.useState(0);
  const [redirect, setRedirect] = React.useState(false);
  const [openRawDataModal, setOpenRawDataModal] = React.useState(false);
  const fetchTransactions = useFetch<{ data: ITransactionDetails }>({
    method: 'get',
    url: `${URLS.TRANSACTION_URL}/${id}`,
  });

  const fetchSummary = useFetch<ISummary>({ method: 'get', url: URLS.SUMMARY_URL });

  const toggleOpenRawData = () => setOpenRawDataModal(prevState => !prevState);
  const [transactionEvents, handleClickSort] = useSortData<TransactionEvent>({
    inititalData: transaction?.transactionEvents || null,
  });
  const handleTransactionFetch = () => {
    fetchTransactions.fetchData().then(response => {
      if (!response?.data) {
        setRedirect(true);
      } else {
        setTransaction(response.data);
      }
    });
  };
  const handleExchangeRateFetch = () => {
    fetchSummary.fetchData().then(response => {
      if (!response) return null;
      return setExchangeRate(response?.currentStats?.usdPrice || 0);
    });
  };

  React.useEffect(() => {
    handleTransactionFetch();
    handleExchangeRateFetch();
  }, [id]);

  const generateTransactionEvents = (
    events: TransactionEvent[] | null,
    type: DirectionType,
  ): RowsProps[] => {
    if (!events) {
      return [];
    }
    const typeTransactionEvents = events.filter(
      transactionEvent => transactionEvent.direction === type,
    );

    const tableTransactionEvents = typeTransactionEvents.map(({ amount, address }, index) => {
      return {
        id: index,
        data: [
          {
            id: 1,
            value: (
              <Grid container alignItems="center" wrap="nowrap">
                <CopyButton copyText={address} />
                <RouterLink
                  route={`${ROUTES.ADDRESS_DETAILS}/${address}`}
                  value={address}
                  textSize="large"
                />
              </Grid>
            ),
          },
          { id: 2, value: formatNumber(amount, { decimalsLength: 2 }) },
          { id: 3, value: formatNumber(amount * exchangeRate, { decimalsLength: 2 }, '$') },
        ],
      };
    });

    return tableTransactionEvents;
  };

  const generateTransactionTable = ({
    blockHash,
    timestamp,
    block,
  }: ITransactionDetails): RowsProps[] => {
    return [
      {
        id: 1,
        data: [
          { id: 1, value: block.confirmations },
          {
            id: 2,
            value:
              blockHash !== 'N/A' ? (
                <RouterLink route={`${ROUTES.BLOCK_DETAILS}/${blockHash}`} value={blockHash} />
              ) : (
                '-'
              ),
          },
          { id: 3, value: formattedDate(timestamp) },
        ],
      },
    ];
  };

  if (redirect) {
    return <Redirect to={ROUTES.NOT_FOUND} />;
  }

  return transaction ? (
    <>
      <Header title="Transaction Details" />
      <Grid container direction="column" spacing={2}>
        <Styles.TransactionDesc item onClick={toggleOpenRawData}>
          {generateTableTitle(transaction)}
        </Styles.TransactionDesc>
        <TransactionRawData
          open={openRawDataModal}
          toggleOpen={toggleOpenRawData}
          rawData={transaction.rawData}
        />
        <Grid item>
          <Table headers={transactionHeaders} rows={generateTransactionTable(transaction)} />
        </Grid>
        <Grid container spacing={6}>
          {transaction.isNonStandard ? (
            generateNonStandardTransactionInfo()
          ) : (
            <>
              <Grid item xs={12} lg={6}>
                {transaction.coinbase === 1 ? (
                  generateCoinbaseInfo(transaction.totalAmount)
                ) : (
                  <Table
                    title="Input Addresses"
                    headers={addressHeaders}
                    handleClickSort={handleClickSort}
                    rows={generateTransactionEvents(transactionEvents, 'Outgoing')}
                  />
                )}
              </Grid>
              <Grid item xs={12} lg={6}>
                <Table
                  title="Recipients"
                  headers={addressHeaders}
                  rows={generateTransactionEvents(transactionEvents, 'Incoming')}
                  handleClickSort={handleClickSort}
                />
              </Grid>
            </>
          )}
        </Grid>
      </Grid>
    </>
  ) : null;
};

export default TransactionDetails;
