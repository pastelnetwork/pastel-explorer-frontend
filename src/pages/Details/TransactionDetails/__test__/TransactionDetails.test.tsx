import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import 'jest-styled-components';

import { MyMockType } from '@utils/types/MockType';
import i18next from '../../../../utils/helpers/i18n';
import TransactionDetails from '../TransactionDetails';

jest.mock('react-router-dom', () => ({
  Link: jest.requireActual('react-router-dom').Link,
  useParams: () => ({
    id: '5c219c1bf6d5a5fe2d368d1b9b7c95786b33e5cf3f86458827c382e3ea4e2453',
  }),
}));
jest.mock('i18next-http-backend');
jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            // noop
          }),
      },
    };
  },
  initReactI18next: {
    type: '3rdParty',
    init: () => {
      // noop
    },
  },
}));
i18next.t = jest.fn().mockImplementation((...arg) => {
  return arg[0];
}) as MyMockType;
const mockStore = configureMockStore();
const store = mockStore({});

describe('pages/TransactionDetails', () => {
  const wrapper = shallow(
    <Provider store={store}>
      <TransactionDetails />
    </Provider>,
  );

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
