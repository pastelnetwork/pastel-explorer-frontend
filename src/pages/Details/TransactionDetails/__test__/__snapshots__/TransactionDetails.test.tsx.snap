// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/TransactionDetails renders correctly 1`] = `ShallowWrapper {}`;
